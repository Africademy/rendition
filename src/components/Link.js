import { h } from 'preact'
import Text from './Text'
import get from 'lodash/get'
import omit from 'lodash/omit'
import { darken } from '../utils'

let Base = Text.extend`
  text-decoration: ${props => props.decor || 'none'};
  cursor: ${props => (props.disabled ? 'not-allowed' : 'pointer')};
  opacity: ${props => (props.disabled ? 0.65 : 1)};
  display: inline-block;

  &:active,
  &:hover {
    color: ${props =>
    !props.disabled &&
      darken(get(props.theme.colors, props.color) || props.color)};
  }
`

const Link = ({ is, blank, ...props }) => {
  if (is) {
    Base = Base.withComponent(is)
  }
  if (props.disabled) {
    props = omit(props, 'href')
  }
  return (
    <Base {...props} rel={blank && 'noopener'} target={blank && '_blank'} />
  )
}

Link.defaultProps = {
  is: 'a',
  color: 'primary.main'
}

export default Link
