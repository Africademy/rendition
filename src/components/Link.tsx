import assign = require('lodash/assign');
import get = require('lodash/get');
import omit = require('lodash/omit');
import * as React from 'react';
import { withProps } from 'recompose';
import styled from 'styled-components';
import asRendition from '../asRendition';
import { DefaultProps, RenditionSystemProps } from '../common-types';
import { darken, monospace } from '../utils';
import { align, bold, caps } from './Txt';

let Base = styled.a<InternalLinkProps>`
  ${align}
  ${monospace as any};
  ${caps as any}
  ${bold as any}

  text-decoration: ${props => props.decor || 'none'};
  cursor: ${props => (props.disabled ? 'not-allowed' : 'pointer')};
  opacity: ${props => (props.disabled ? 0.65 : 1)};
  display: inline-block;

  &:active,
  &:hover {
    color: ${props =>
			!props.disabled &&
			(darken(
				get(props.theme.colors, props.color as any) || props.color,
			) as any)};
  }
`;

const Link = ({ is, blank, children, ...props }: InternalLinkProps) => {
	if (is) {
		Base = Base.withComponent(is as any);
	}
	if (props.disabled) {
		props = omit(props, 'href');
	}
	return (
		<Base
			{...props}
			rel={blank ? 'noopener' : undefined}
			target={blank ? '_blank' : undefined}
		>
			{children || props.href}
		</Base>
	);
};

const setDefaultProps = withProps((props: InternalLinkProps) => {
	return assign(
		{
			color: `primary.main`,
		},
		props,
	);
});

export interface InternalLinkProps extends DefaultProps {
	blank?: boolean;
	disabled?: boolean;
	download?: any;
	href?: string;
	hrefLang?: string;
	media?: string;
	rel?: string;
	target?: string;
	type?: string;
	is?: string;
	decor?: string;
	color?: string;
}

export type LinkProps = InternalLinkProps & RenditionSystemProps;

export default asRendition<React.FunctionComponent<LinkProps>>(
	Link,
	[setDefaultProps],
	['color'],
);
