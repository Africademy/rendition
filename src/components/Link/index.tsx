import get from 'lodash/get';
import omit from 'lodash/omit';
import * as React from 'react';
import { withProps } from 'recompose';
import styled from 'styled-components';
import asRendition from '../../asRendition';
import { DefaultProps, RenditionSystemProps } from '../../common-types';
import { darken, monospace, isExternalUrl, toRelativeUrl } from '../../utils';
import path from 'path';
import { align, bold, caps } from '../Txt';
import { withRouter } from 'react-router-dom';

const Base = styled.a<InternalLinkProps>`
  ${align}
  ${monospace};
  ${caps}
  ${bold}

  text-decoration: ${(props) => props.decor || 'none'};
  cursor: ${(props) => (props.disabled ? 'not-allowed' : 'pointer')};
  opacity: ${(props) => (props.disabled ? 0.65 : 1)};
  display: inline-block;

  &:active,
  &:hover {
    color: ${(props) =>
			!props.disabled &&
			(darken(
				get(props.theme.colors, props.color as any) || props.color,
			) as any)};
  }
`;

export class RouterLink extends React.Component {
	constructor(props) {
		super(props);

		this.navigate = this.navigate.bind(this);
	}

	makeUrl() {
		const { append, location, to } = this.props;

		if (to) {
			return toRelativeUrl(to);
		}

		if (append) {
			return path.join(location.pathname, append);
		}

		return null;
	}

	navigate(event) {
		const { blank, to } = this.props;

		// If control or meta keys are pressed use default browser behaviour
		if (event.ctrlKey || event.metaKey) {
			return true;
		}

		// If the link is external use default browser behaviour
		if (blank || isExternalUrl(to)) {
			return true;
		}

		event.preventDefault();
		const { history } = this.props;

		const url = this.makeUrl();

		history.push(url);
		return false;
	}

	render() {
		const props = omit(this.props, [
			'match',
			'location',
			'history',
			'to',
			'append',
		]);

		const url = this.makeUrl();

		return (
			<Link
				{...props}
				href={url}
				// We should only navigate when `url` is defined
				onClick={url ? this.navigate : props.onClick}
			/>
		);
	}
}

const Link = ({ is, blank, children, ...props }: InternalLinkProps) => {
	if (props.disabled) {
		props = omit(props, 'href');
	}
	return (
		<Base
			{...props}
			as={is}
			rel={blank ? 'noopener' : undefined}
			target={blank ? '_blank' : undefined}
		>
			{children || props.href}
		</Base>
	);
};

const setDefaultProps = withProps((props: InternalLinkProps) => {
	return Object.assign(
		{
			color: `primary.main`,
		},
		props,
	);
});

export interface InternalLinkProps extends DefaultProps {
	blank?: boolean;
	disabled?: boolean;
	download?: any;
	href?: string;
	hrefLang?: string;
	media?: string;
	rel?: string;
	target?: string;
	type?: string;
	is?: React.ElementType;
	decor?: string;
	color?: string;
}

export type LinkProps = InternalLinkProps & RenditionSystemProps;

export default asRendition<React.Component<LinkProps>>(
	withRouter(RouterLink),
	[setDefaultProps],
	['color'],
);
