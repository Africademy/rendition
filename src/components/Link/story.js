import * as React from 'react'
import { storiesOf } from '@storybook/react'
import withReadme from 'storybook-readme/with-readme'
import { Box, Link, Provider } from '../../'
import Readme from './README.md'

storiesOf('Core/Link', module)
  .addDecorator(withReadme(Readme))
  .add('Standard', () => {
    return (
      <Provider>
        <Box m={3}>
          <div>
            <Link href={`#`}>Internal Link</Link>
          </div>
          <div>
            <Link href={`https://balena.io`} blank>
              External Link
            </Link>
          </div>
          <div>
            <Link disabled href={`https://balena.io`} blank>
              Disabled Link
            </Link>
          </div>
        </Box>
      </Provider>
    )
  })
