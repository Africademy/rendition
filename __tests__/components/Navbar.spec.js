/* globals expect, test, describe */
import React from 'react'
import renderer from 'react-test-renderer'
import Link from '../../src/components/Link'
import {
  Navbar,
  Provider
} from '../../dist'

describe('Navbar renders correctly', () => {
  test('with more than one children prop', () => {
    const component = renderer.create(
      <Provider>
        <Navbar color='white'>
          <Link color='white' href={'/docs/'}>
            Docs
          </Link>
          <Link color='white' href={'/changelog/'}>
            changelog
          </Link>
          <Link color='white' href={'/gitter/'}>
            gitter
          </Link>
        </Navbar>
      </Provider>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('with a single children prop', () => {
    const component = renderer.create(
      <Provider>
        <Navbar color='white'>
          <Link color='white' href={'/docs/'}>
            Docs
          </Link>
        </Navbar>
      </Provider>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('with no children prop', () => {
    const component = renderer.create(
      <Provider>
        <Navbar color='white' />
      </Provider>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
