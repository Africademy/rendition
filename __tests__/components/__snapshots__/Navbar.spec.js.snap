// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders correctly with a single children prop 1`] = `
<div
  className="sc-iyvyFf gFhKYN StyledGrommet-sc-19lkkz7-0 bBubtL"
>
  <div
    className="sc-chPdSV GFDdw sc-kAzzGY dLnIPy"
  >
    <div
      className="sc-hZSUBg eoZWeN sc-bsbRJL tKRUY"
    >
      <div
        className="sc-kgoBCf kqLfuS sc-chPdSV jpfrZo sc-kAzzGY dLnIPy"
      >
        <div
          className="sc-kgoBCf cRCsAq sc-chPdSV jpfrZo sc-kAzzGY dLnIPy"
        >
          <div
            className="sc-cMhqgX kOxYlb sc-chPdSV gwIGaM sc-kAzzGY dLnIPy"
          />
          <div
            className="sc-iuJeZd bLYkWY sc-chPdSV yrOiB sc-kAzzGY dLnIPy"
            display={
              Array [
                "block",
                "none",
              ]
            }
            onClick={[Function]}
          >
            <svg
              fill="currentColor"
              height="1em"
              preserveAspectRatio="xMidYMid meet"
              style={
                Object {
                  "color": undefined,
                  "verticalAlign": "middle",
                }
              }
              viewBox="0 0 40 40"
              width="1em"
            >
              <g>
                <path
                  d="m3.4 9.1h33.2v3.4h-33.2v-3.4z m0 8.4h33.2v3.4h-33.2v-3.4z m0 8.4h33.2v3.2h-33.2v-3.2z"
                />
              </g>
            </svg>
          </div>
        </div>
        <div
          className="sc-kgoBCf sc-esOvli jSeSux sc-chPdSV jgcBzP sc-kAzzGY dLnIPy"
        >
          <div
            className="sc-kgoBCf kqLfuS sc-chPdSV kDIXHv sc-kAzzGY dLnIPy"
          >
            <div
              className="sc-chPdSV kDQfPd sc-kAzzGY dLnIPy"
              onClick={[Function]}
            >
              <a
                className="sc-gZMcBi krgeid sc-iwsKbI jkCyoU"
                color="white"
                href="/docs/"
              >
                Docs
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Navbar renders correctly with more than one children prop 1`] = `
<div
  className="sc-iyvyFf gFhKYN StyledGrommet-sc-19lkkz7-0 bBubtL"
>
  <div
    className="sc-chPdSV GFDdw sc-kAzzGY dLnIPy"
  >
    <div
      className="sc-hZSUBg eoZWeN sc-bsbRJL tKRUY"
    >
      <div
        className="sc-kgoBCf kqLfuS sc-chPdSV jpfrZo sc-kAzzGY dLnIPy"
      >
        <div
          className="sc-kgoBCf cRCsAq sc-chPdSV jpfrZo sc-kAzzGY dLnIPy"
        >
          <div
            className="sc-cMhqgX kOxYlb sc-chPdSV gwIGaM sc-kAzzGY dLnIPy"
          />
          <div
            className="sc-iuJeZd bLYkWY sc-chPdSV yrOiB sc-kAzzGY dLnIPy"
            display={
              Array [
                "block",
                "none",
              ]
            }
            onClick={[Function]}
          >
            <svg
              fill="currentColor"
              height="1em"
              preserveAspectRatio="xMidYMid meet"
              style={
                Object {
                  "color": undefined,
                  "verticalAlign": "middle",
                }
              }
              viewBox="0 0 40 40"
              width="1em"
            >
              <g>
                <path
                  d="m3.4 9.1h33.2v3.4h-33.2v-3.4z m0 8.4h33.2v3.4h-33.2v-3.4z m0 8.4h33.2v3.2h-33.2v-3.2z"
                />
              </g>
            </svg>
          </div>
        </div>
        <div
          className="sc-kgoBCf sc-esOvli jSeSux sc-chPdSV jgcBzP sc-kAzzGY dLnIPy"
        >
          <div
            className="sc-kgoBCf kqLfuS sc-chPdSV kDIXHv sc-kAzzGY dLnIPy"
          >
            <div
              className="sc-chPdSV kDQfPd sc-kAzzGY dLnIPy"
              onClick={[Function]}
            >
              <a
                className="sc-gZMcBi krgeid sc-iwsKbI jkCyoU"
                color="white"
                href="/docs/"
              >
                Docs
              </a>
            </div>
            <div
              className="sc-chPdSV kDQfPd sc-kAzzGY dLnIPy"
              onClick={[Function]}
            >
              <a
                className="sc-gZMcBi krgeid sc-iwsKbI jkCyoU"
                color="white"
                href="/changelog/"
              >
                changelog
              </a>
            </div>
            <div
              className="sc-chPdSV kDQfPd sc-kAzzGY dLnIPy"
              onClick={[Function]}
            >
              <a
                className="sc-gZMcBi krgeid sc-iwsKbI jkCyoU"
                color="white"
                href="/gitter/"
              >
                gitter
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Navbar renders correctly with no children prop 1`] = `
<div
  className="sc-iyvyFf gFhKYN StyledGrommet-sc-19lkkz7-0 bBubtL"
>
  <div
    className="sc-chPdSV GFDdw sc-kAzzGY dLnIPy"
  >
    <div
      className="sc-hZSUBg eoZWeN sc-bsbRJL tKRUY"
    >
      <div
        className="sc-kgoBCf kqLfuS sc-chPdSV jpfrZo sc-kAzzGY dLnIPy"
      >
        <div
          className="sc-kgoBCf cRCsAq sc-chPdSV jpfrZo sc-kAzzGY dLnIPy"
        >
          <div
            className="sc-cMhqgX kOxYlb sc-chPdSV gwIGaM sc-kAzzGY dLnIPy"
          />
          <div
            className="sc-iuJeZd bLYkWY sc-chPdSV yrOiB sc-kAzzGY dLnIPy"
            display={
              Array [
                "block",
                "none",
              ]
            }
            onClick={[Function]}
          >
            <svg
              fill="currentColor"
              height="1em"
              preserveAspectRatio="xMidYMid meet"
              style={
                Object {
                  "color": undefined,
                  "verticalAlign": "middle",
                }
              }
              viewBox="0 0 40 40"
              width="1em"
            >
              <g>
                <path
                  d="m3.4 9.1h33.2v3.4h-33.2v-3.4z m0 8.4h33.2v3.4h-33.2v-3.4z m0 8.4h33.2v3.2h-33.2v-3.2z"
                />
              </g>
            </svg>
          </div>
        </div>
        <div
          className="sc-kgoBCf sc-esOvli jSeSux sc-chPdSV jgcBzP sc-kAzzGY dLnIPy"
        />
      </div>
    </div>
  </div>
</div>
`;
